Compute bindings factors and group commitment

pdyraga opened the pull request:
This PR introduces `computeBindingFactors` and `computeGroupCommitment` based on @eth-r's prototype implementation. The code is almost 1:1 as in the prototype with the exception that we use a map to represent binding factors. This way we can eliminate `binding_factor_for_participant` search.

The `Deserialize` function of the elliptic curve point BIP-340 implementation validates now if the point is on the curve. The validation of group commitments has been moved to the beginning of round two.

Worth noting, `computeBindingFactors` and `computeGroupCommitment` lack unit tests. It is hard to come up with unit tests that wouldn't replicate implementation or work on hardcoded numbers that don't make a lot of sense alone, out of context. I am planning to cover round two in unit tests as a whole piece.

pdyraga committed:
Validate the point when deserializing

When deserializing a point we check if it is a valid point that lies on
the curve.

pdyraga committed:
Call validateGroupCommitments at the beginning of Round2

There will be more functions than just encodeGroupCommitments operating
on the commitments so it makes sense to validate them at the very
beginning.

No logic in encodeGroupCommitment was altered as a part of this commit
other than the call to validateGroupCommitments was removed from this
function as it is a part of Round2 function now.

pdyraga committed:
FROST function to compute binding factors in round 2 of the protocol

The function calling computeBindingFactors must ensure a valid number of
commitments have been received and call validateGroupCommitment to validate
the received commitments.

pdyraga committed:
FROST function to compute group commitment in round 2 of the protocol

computeGroupCommitment implements def compute_group_commitment(commitment_list,
binding_factor_list) function from [FROST], as defined in section 4.5. Group
Commitment Computation.

pdyraga committed:
Added missing [FROST] pseudocode reference

eth-r commented on frost/bip340.go:
```diff
@@ -61,18 +83,22 @@ func (b *Bip340Curve) SerializePoint(p *Point) []byte {
-	return &Point{x, y}
+	point := &Point{x, y}
+
+	if !b.IsNotIdentity(point) {
```
The naming of this function is a bit confusing in context, because we're also checking that it's generally a valid point, not just that it's not the point at infinity. Should we leave a clarifying comment or rename it?

pdyraga replied:
Agreed, this naming is unfortunate. Please see ff1b58dc0e90afe4bfd8affc80328e17845e1d77.

eth-r commented on frost/bip340_test.go:
```diff
@@ -57,6 +57,9 @@ func TestBip340CurveDeserialize(t *testing.T) {
 			input: append(serialized, 0x1),
 		},
+		"not on the curve": {
+			input: curve.SerializePoint(&Point{big.NewInt(1), big.NewInt(2)}),
```
Should we also test it on the infinity point?

pdyraga replied:
Good point. Please see a7c405d30f15f800a7e0ff4f0a9186608e05a055.

pdyraga committed:
Renamed IsNotIdentity to IsPointOnCurve

For elliptic curve cryptography, we do not have an identity element but we take
the point at infinity as an identity element. The point at infinity is an extra
point O and it is not on the curve. In our case it is enough to validate if the
point we are deserializing is on the curve and we can make it clear by renaming
the function.

pdyraga committed:
Improved test coverage for curve operations

Added basic unit tests for EcBaseMul, EcMul, and EcAdd.
Included identity element in DeserializePoint test cases.

eth-r reviewed (approved):
Looks good now.
